libs/pika/async_mpi/src/mpi_future.cpp

```diff
@@ -226,150 +357,95 @@ namespace pika { namespace mpi { namespace experimental {
             static auto poll_deb =
                 mpi_debug.make_timer(1, debug::str<>("Poll - lock success"));
             // output mpi debug info every N seconds
-            mpi_debug.timed(poll_deb, detail::get_mpi_info());
+            mpi_debug.timed(poll_deb, detail::mpi_data_);
         }
 
+        // Before moving requests from the queue to the vector
+        compact_vectors();
+
+        // Move requests in the queue (that have not yet been polled for)
+        // into the polling vector ...
+        // Number in_flight does not change during this section as one
+        // is moved off the queue and into the vector
+
+        detail::request_callback req_callback;
+        while (
+            detail::mpi_data_.request_callback_queue_.try_dequeue(req_callback))
         {
-            // have any requests been made that need to be handled?
-            detail::request_callback req_callback;
-            while (
-                detail::get_request_callback_queue().try_dequeue(req_callback))
-            {
-                --(detail::get_mpi_info().requests_queue_size_);
-                add_to_request_callback_vector(PIKA_MOVE(req_callback));
-            }
+            --detail::mpi_data_.request_queue_size_;
+            add_to_request_callback_vector(PIKA_MOVE(req_callback));
         }
 
-        bool keep_trying = !requests_vector.empty();
-        while (keep_trying)
-        {
-            int index = 0;
-            int flag = false;
-            MPI_Status status;
+        int outcount = 0;
+        int vsize = detail::mpi_data_.request_vector_.size();
+        detail::mpi_data_.indices_vector_.resize(vsize);
+        detail::mpi_data_.status_vector_.resize(vsize);
+        int result =
+            MPI_Testsome(vsize, detail::mpi_data_.request_vector_.data(),
+                &outcount, detail::mpi_data_.indices_vector_.data(),
+                detail::mpi_data_.status_vector_.data());
 
-            int result = MPI_Testany(static_cast<int>(requests_vector.size()),
-                requests_vector.data(), &index, &flag, &status);
+        if (result != MPI_SUCCESS)
+            std::terminate();
 
-            if constexpr (mpi_debug.is_enabled())
+        if constexpr (mpi_debug.is_enabled())
+        {
+            // output a heartbeat every seconds
+            static auto poll_deb =
+                mpi_debug.make_timer(1, debug::str<>("Poll - success"));
+            mpi_debug.timed(poll_deb, detail::mpi_data_, "outcount", outcount,
+                debug::dec<4>(outcount));
+        }
+
+        // for each completed request
+        for (int i = 0; i < outcount; ++i)
+        {
+            size_t index = detail::mpi_data_.indices_vector_[i];
+            if (detail::mpi_data_.status_vector_[i].MPI_ERROR != MPI_SUCCESS)
```

Comment:
I'd remove this and forward the status to the callback.

Reply:
done